packages/instantsearch.js/src/widgets/index/index.ts

```diff
@@ -277,7 +279,14 @@ const index = (widgetParams: IndexWidgetParams): IndexWidget => {
       return localParent;
     },
 
-    createURL(nextState: SearchParameters) {
+    createURL(
+      nextState: SearchParameters | ((state: IndexUiState) => IndexUiState)
+    ) {
+      if (typeof nextState === 'function') {
+        return localInstantSearchInstance!._createURL({
+          [indexId]: nextState(localUiState),
+        });
+      }
       return localInstantSearchInstance!._createURL({
         [indexId]: getLocalWidgetsUiState(localWidgets, {
           searchParameters: nextState,
```

Comment:
this is the real meat of the PR. Because you pass a createURL with only uiState, we don't need to calculate the UiState based on the localWidgets (which aren't yet mounted) and can pass the UiState directly to _createURL